File: java/bypstest/src/com/wilutions/byps/test/TestRemoteListTypes.java
package com.wilutions.byps.test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.TreeSet;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import com.wilutions.byps.BException;
import com.wilutions.byps.test.api.BClient_Testser;
import com.wilutions.byps.test.api.list.ListListTypes;
import com.wilutions.byps.test.api.list.ListTypes;
import com.wilutions.byps.test.api.prim.PrimitiveTypes;
import com.wilutions.byps.test.api.remote.RemoteListTypes;

/**
 * Tests for interface functions with list types.
 * This test requires the webapp bypstest-srv to be started. 
 * 
 *
 */
public class TestRemoteListTypes {
	
	BClient_Testser client;
	RemoteListTypes remote;
	private Log log = LogFactory.getLog(TestRemoteListTypes.class);

	@Before
	public void setUp() throws BException, InterruptedException {
		client = TestUtilsHttp.createClient();
		remote = client.remoteListTypes;
	}
	
	@After
	public void tearDown() {
		if (client != null) {
			client.done();
		}
	}
	
	@Test
	public void testRemoteListTypes() throws BException, InterruptedException {
		log.info("testRemoteListTypes(");
		
		ListTypes obj = new ListTypes();
		obj.boolean1 = new ArrayList<Boolean>(Arrays.asList(true, false));
		obj.byte1 = new ArrayList<Byte>(Arrays.asList((byte)1,(byte)2,(byte)3));
		obj.char1 = new ArrayList<Character>(Arrays.asList('a', 'b'));
		obj.double1 = new ArrayList<Double>(Arrays.asList(0.0, 0.1, 0.2));
		obj.float1 = new ArrayList<Float>(Arrays.asList(1f, 2f, 3f, 4f));
		obj.int1 = new ArrayList<Integer>(Arrays.asList(1, 2, 3, 4, 5));
		obj.long1 = new ArrayList<Long>(Arrays.asList(1L, 2L, 3L, 4L, 5L , 6L));
		obj.primitiveTypes1 = new ArrayList<PrimitiveTypes>(Arrays.asList(new PrimitiveTypes[] { TestUtils.createObjectPrimitiveTypes() }));
		obj.short1 = new ArrayList<Short>(Arrays.asList((short)1));
		obj.string1 = new ArrayList<String>(Arrays.asList("a", "b", "c"));
		obj.obj1 = new ArrayList<Object>();
		obj.obj1.add(TestUtils.createObjectPrimitiveTypes());
		
		remote.setBoolean1(obj.boolean1);
		TestUtils.assertEquals(log, "boolean1", obj.boolean1, remote.getBoolean1());
		remote.setByte1(obj.byte1);
		TestUtils.assertEquals(log, "byte1", obj.byte1, remote.getByte1());
		remote.setChar1(obj.char1);
		TestUtils.assertEquals(log, "char1", obj.char1, remote.getChar1());
		remote.setDouble1(obj.double1);
		TestUtils.assertEquals(log, "double1", obj.double1, remote.getDouble1());
		remote.setFloat1(obj.float1);
		TestUtils.assertEquals(log, "float1", obj.float1, remote.getFloat1());
		remote.setInt1(obj.int1);
		TestUtils.assertEquals(log, "int1", obj.int1, remote.getInt1());
		remote.setLong1(obj.long1);
		TestUtils.assertEquals(log, "long1", obj.long1, remote.getLong1());
		remote.setPrimitiveTypes1(obj.primitiveTypes1);
		TestUtils.assertEquals(log, "primitiveTypes1", obj.primitiveTypes1, remote.getPrimitiveTypes1());
		remote.setShort1(obj.short1);
		TestUtils.assertEquals(log, "short1", obj.short1, remote.getShort1());
		remote.setString1(obj.string1);
		TestUtils.assertEquals(log, "string1", obj.string1, remote.getString1());
		remote.setObj1(obj.obj1);
		TestUtils.assertEquals(log, "obj1", obj.obj1, remote.getObj1());
		
		log.info(")testRemoteListTypes");
	}
	
	@Test
	public void testRemoteListMapSetListInteger() throws BException, InterruptedException {
		log.info("testRemoteListMapSetListInteger(");
		
		ListListTypes obj = new ListListTypes();
		List<Map<Integer,List<TreeSet<Integer>>>> list = new ArrayList<Map<Integer,List<TreeSet<Integer>>>>();
		for (int i = 0; i < 2; i++) {
			Map<Integer,List<TreeSet<Integer>>> map = new TreeMap<Integer,List<TreeSet<Integer>>>();
			for (int j = 0; j < 3; j++) {
				List<TreeSet<Integer>> set = new ArrayList<TreeSet<Integer>>();
				for (int k = 0; k < 4; k++) {
					TreeSet<Integer> list2 = new TreeSet<Integer>();
					for (int n = 0; n < 5; n++) {
						Integer pt = (i+1) * (j+1) * (k+1) * (n+1);
						list2.add(pt);
					}
					set.add(list2);
				}
				map.put(j, set);
			}
			list.add(map);
		}
		obj.int3 = list;
		
		remote.setInt4(obj.int3);
		TestUtils.assertEquals(log, "int4", obj.int3, remote.getInt4());
		
		log.info(")testRemoteListMapSetListInteger");
	}

	@Test
	public void testRemoteListListInteger() throws BException, InterruptedException {
		log.info("testRemoteListListInteger(");
		
		ListListTypes obj = new ListListTypes();
		List<List<Integer>> list = new ArrayList<List<Integer>>();
		for (int i = 0; i < 3; i++) {
			ArrayList<Integer> itemList = new ArrayList<Integer>();
			for (int j = 0; j < i+1; j++) {
				itemList.add(j);
			}
			list.add(itemList);
		}
		obj.int2 = list;
		
		remote.setInt3(obj.int2);
		TestUtils.assertEquals(log, "int3", obj.int2, remote.getInt3());
		
		log.info(")testRemoteListListInteger");
	}

}

File: csharp/bypstest/TestRemoteListTypes.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using com.wilutions.byps;
using com.wilutions.byps.test.api;
using com.wilutions.byps.test.api.remote;
using com.wilutions.byps.test.api.list;
using com.wilutions.byps.test.api.prim;
using System.Collections.Generic;

namespace bypstest
{
    [TestClass]
    public class TestRemoteListTypes
    {
        Log log = LogFactory.getLog(typeof(TestRemoteMapTypes));
        BClient_Testser client;
        RemoteListTypes remote;

        [TestInitialize]
        public void setUp()
        {
            client = TestUtilsHttp.createClient(BWireFlags.GZIP);
            remote = client.RemoteListTypes;
        }

        [TestCleanup]
        public void tearDown()
        {
            client.done();
        }

 	    [TestMethod]
	    public void testRemoteListTypes() {
		    log.info("testRemoteListTypes(");
		
		    ListTypes obj = new ListTypes();
            obj.Boolean1 = new List<bool>(new bool[] { true, false });
		    obj.Byte1 = new List<byte>(new byte[] {(byte)1,(byte)2,(byte)3 });
		    obj.Char1 = new List<char>(new char [] {'a', 'b'});
            obj.Double1 = new List<double>(new double[] { 0.0, 0.1, 0.2 });
            obj.Float1 = new List<float>(new float[] { 1f, 2f, 3f, 4f });
            obj.Int1 = new List<int>(new int[] { 1, 2, 3, 4, 5 });
            obj.Long1 = new List<long>(new long[] { 1L, 2L, 3L, 4L, 5L, 6L });
		    obj.PrimitiveTypes1 = new List<PrimitiveTypes>(new PrimitiveTypes[] { TestUtils.createObjectPrimitiveTypes() });
		    obj.Short1 = new List<short>(new short[] {(short)1});
		    obj.String1 = new List<String>(new String[] {"a", "b", "c"});
		    obj.Obj1 = new List<Object>();
		    obj.Obj1.Add(TestUtils.createObjectPrimitiveTypes());
		
		    remote.SetBoolean1(obj.Boolean1);
		    TestUtils.assertEquals(log, "boolean1", obj.Boolean1, remote.GetBoolean1());
		    remote.SetByte1(obj.Byte1);
		    TestUtils.assertEquals(log, "byte1", obj.Byte1, remote.GetByte1());
		    remote.SetChar1(obj.Char1);
		    TestUtils.assertEquals(log, "char1", obj.Char1, remote.GetChar1());
		    remote.SetDouble1(obj.Double1);
		    TestUtils.assertEquals(log, "double1", obj.Double1, remote.GetDouble1());
		    remote.SetFloat1(obj.Float1);
		    TestUtils.assertEquals(log, "float1", obj.Float1, remote.GetFloat1());
		    remote.SetInt1(obj.Int1);
		    TestUtils.assertEquals(log, "int1", obj.Int1, remote.GetInt1());
		    remote.SetLong1(obj.Long1);
		    TestUtils.assertEquals(log, "long1", obj.Long1, remote.GetLong1());
		    remote.SetPrimitiveTypes1(obj.PrimitiveTypes1);
		    TestUtils.assertEquals(log, "primitiveTypes1", obj.PrimitiveTypes1, remote.GetPrimitiveTypes1());
		    remote.SetShort1(obj.Short1);
		    TestUtils.assertEquals(log, "short1", obj.Short1, remote.GetShort1());
		    remote.SetString1(obj.String1);
		    TestUtils.assertEquals(log, "string1", obj.String1, remote.GetString1());
		    remote.SetObj1(obj.Obj1);
		    TestUtils.assertEquals(log, "obj1", obj.Obj1, remote.GetObj1());
		
		    log.info(")testRemoteListTypes");
	    }
	
	    [TestMethod]
	    public void testRemoteListMapSetListInteger() {
		    log.info("testRemoteListMapSetListInteger(");
		
		    ListListTypes obj = new ListListTypes();
            List<IDictionary<int, IList<HashSet<int>>>> list = new List<IDictionary<int, IList<HashSet<int>>>>();
		    for (int i = 0; i < 2; i++) {
			    IDictionary<int,IList<HashSet<int>>> map = new Dictionary<int,IList<HashSet<int>>>();
			    for (int j = 0; j < 3; j++) {
                    IList<HashSet<int>> list2 = new List<HashSet<int>>();
                    for (int k = 0; k < 4; k++)
                    {
                        HashSet<int> set = new HashSet<int>();
                        list2.Add(set);
					    for (int n = 0; n < 5; n++) {
						    int pt = n * k * j * i;
                            set.Add(pt);
					    }
					    
				    }
				    map[j] = list2;
			    }
			    list.Add(map);
		    }
		    obj.Int3 = list;
		
		    remote.SetInt4(obj.Int3);
		    TestUtils.assertEquals(log, "int4", obj.Int3, remote.GetInt4());
		
		    log.info(")testRemoteListMapSetListint");
	    }

	    [TestMethod]
	    public void testRemoteListListint()  {
		    log.info("testRemoteListListint(");
		
		    ListListTypes obj = new ListListTypes();
		    List<IList<int>> list = new List<IList<int>>();
		    for (int i = 0; i < 3; i++) {
			    List<int> itemList = new List<int>();
			    for (int j = 0; j < i+1; j++) {
				    itemList.Add(j);
			    }
		    }
		    obj.Int2 = list;
		
		    remote.SetInt3(obj.Int2);
		    TestUtils.assertEquals(log, "int3", obj.Int2, remote.GetInt3());
		
		    log.info(")testRemoteListListint");
	    }
    }
}
